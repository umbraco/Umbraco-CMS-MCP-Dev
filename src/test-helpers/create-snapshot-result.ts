import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
