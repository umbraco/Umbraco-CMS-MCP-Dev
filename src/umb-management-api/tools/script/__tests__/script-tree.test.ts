import GetScriptTreeRootTool from "../get/get-script-tree-root.js";
import GetScriptTreeChildrenTool from "../get/get-script-tree-children.js";
import GetScriptTreeAncestorsTool from "../get/get-script-tree-ancestors.js";
import { ScriptBuilder } from "./helpers/script-builder.js";
import { ScriptFolderBuilder } from "./helpers/script-folder-builder.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { jest } from "@jest/globals";

const ROOT_FOLDER_NAME = "_RootFolder";
const ROOT_SCRIPT_NAME = "_RootScript";
const ROOT_SCRIPT_CONTENT = "console.log('root script');";
const CHILD_FOLDER_NAME = "_ChildFolder";
const CHILD_SCRIPT_NAME = "_ChildScript";
const CHILD_SCRIPT_CONTENT = "console.log('child script');";
const GRANDCHILD_SCRIPT_NAME = "_GrandchildScript";
const GRANDCHILD_SCRIPT_CONTENT = "console.log('grandchild script');";

describe("script-tree", () => {
  let originalConsoleError: typeof console.error;
  let rootFolderBuilder: ScriptFolderBuilder;
  let rootScriptBuilder: ScriptBuilder;
  let childFolderBuilder: ScriptFolderBuilder;
  let childScriptBuilder: ScriptBuilder;
  let grandchildScriptBuilder: ScriptBuilder;

  beforeAll(async () => {
    // Create the script tree structure
    rootFolderBuilder = new ScriptFolderBuilder();
    rootScriptBuilder = new ScriptBuilder();
    childFolderBuilder = new ScriptFolderBuilder();
    childScriptBuilder = new ScriptBuilder();
    grandchildScriptBuilder = new ScriptBuilder();

    // Create root folder
    await rootFolderBuilder
      .withName(ROOT_FOLDER_NAME)
      .create();

    // Create root script
    await rootScriptBuilder
      .withName(ROOT_SCRIPT_NAME)
      .withContent(ROOT_SCRIPT_CONTENT)
      .create();

    // Create child folder under root folder
    await childFolderBuilder
      .withName(CHILD_FOLDER_NAME)
      .withParent(rootFolderBuilder.getPath())
      .create();

    // Create child script under root folder
    await childScriptBuilder
      .withName(CHILD_SCRIPT_NAME)
      .withContent(CHILD_SCRIPT_CONTENT)
      .withParent(rootFolderBuilder.getPath())
      .create();

    // Create grandchild script under child folder
    await grandchildScriptBuilder
      .withName(GRANDCHILD_SCRIPT_NAME)
      .withContent(GRANDCHILD_SCRIPT_CONTENT)
      .withParent(childFolderBuilder.getPath())
      .create();
  });

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
  });

  afterEach(() => {
    console.error = originalConsoleError;
  });

  afterAll(async () => {
    // Clean up all test script items in reverse order
    await grandchildScriptBuilder.cleanup();
    await childScriptBuilder.cleanup();
    await childFolderBuilder.cleanup();
    await rootScriptBuilder.cleanup();
    await rootFolderBuilder.cleanup();
  });

  describe("get-script-tree-root", () => {
    it("should get root level script items", async () => {
      const result = await GetScriptTreeRootTool().handler(
        {
          take: 100,
        },
        { signal: new AbortController().signal }
      );
      expect(createSnapshotResult(result)).toMatchSnapshot();
    });
  });

  describe("get-script-tree-children", () => {
    it("should get children of root folder", async () => {
      const result = await GetScriptTreeChildrenTool().handler(
        {
          parentPath: rootFolderBuilder.getPath(),
          take: 100,
        },
        { signal: new AbortController().signal }
      );
      expect(createSnapshotResult(result)).toMatchSnapshot();
    });

    it("should get children of child folder", async () => {
      const result = await GetScriptTreeChildrenTool().handler(
        {
          parentPath: childFolderBuilder.getPath(),
          take: 100,
        },
        { signal: new AbortController().signal }
      );
      expect(createSnapshotResult(result)).toMatchSnapshot();
    });

    it("should get empty children for non-folder script", async () => {
      const result = await GetScriptTreeChildrenTool().handler(
        {
          parentPath: grandchildScriptBuilder.getPath(),
          take: 100,
        },
        { signal: new AbortController().signal }
      );
      expect(createSnapshotResult(result)).toMatchSnapshot();
    });
  });

  describe("get-script-tree-ancestors", () => {
    it("should get ancestors of grandchild script", async () => {
      const result = await GetScriptTreeAncestorsTool().handler(
        {
          descendantPath: grandchildScriptBuilder.getPath(),
        },
        { signal: new AbortController().signal }
      );
      expect(createSnapshotResult(result)).toMatchSnapshot();
    });

    it("should get ancestors of child script", async () => {
      const result = await GetScriptTreeAncestorsTool().handler(
        {
          descendantPath: childScriptBuilder.getPath(),
        },
        { signal: new AbortController().signal }
      );
      expect(createSnapshotResult(result)).toMatchSnapshot();
    });

    it("should get empty ancestors for root script", async () => {
      const result = await GetScriptTreeAncestorsTool().handler(
        {
          descendantPath: rootScriptBuilder.getPath(),
        },
        { signal: new AbortController().signal }
      );
      expect(createSnapshotResult(result)).toMatchSnapshot();
    });
  });
});