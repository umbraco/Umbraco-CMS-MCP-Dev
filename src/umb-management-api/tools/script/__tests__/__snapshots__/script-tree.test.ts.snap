// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`script-tree get-script-tree-ancestors should get ancestors of child script 1`] = `
{
  "content": [
    {
      "text": "[{"name":"","path":"/","parent":null,"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_RootFolder","path":"/_RootFolder","parent":{"path":"/","id":"00000000-0000-0000-0000-000000000000"},"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_ChildScript.js","path":"/_RootFolder/_ChildScript.js","parent":{"path":"/_RootFolder","id":"00000000-0000-0000-0000-000000000000"},"isFolder":false,"hasChildren":false,"id":"00000000-0000-0000-0000-000000000000"}]",
      "type": "text",
    },
  ],
}
`;

exports[`script-tree get-script-tree-ancestors should get ancestors of grandchild script 1`] = `
{
  "content": [
    {
      "text": "[{"name":"","path":"/","parent":null,"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_RootFolder","path":"/_RootFolder","parent":{"path":"/","id":"00000000-0000-0000-0000-000000000000"},"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_ChildFolder","path":"/_RootFolder/_ChildFolder","parent":{"path":"/_RootFolder","id":"00000000-0000-0000-0000-000000000000"},"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_GrandchildScript.js","path":"/_RootFolder/_ChildFolder/_GrandchildScript.js","parent":{"path":"/_RootFolder/_ChildFolder","id":"00000000-0000-0000-0000-000000000000"},"isFolder":false,"hasChildren":false,"id":"00000000-0000-0000-0000-000000000000"}]",
      "type": "text",
    },
  ],
}
`;

exports[`script-tree get-script-tree-ancestors should get empty ancestors for root script 1`] = `
{
  "content": [
    {
      "text": "[{"name":"","path":"/","parent":null,"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_RootScript.js","path":"/_RootScript.js","parent":{"path":"/","id":"00000000-0000-0000-0000-000000000000"},"isFolder":false,"hasChildren":false,"id":"00000000-0000-0000-0000-000000000000"}]",
      "type": "text",
    },
  ],
}
`;

exports[`script-tree get-script-tree-children should get children of child folder 1`] = `
{
  "content": [
    {
      "text": "{"total":1,"items":[{"name":"_GrandchildScript.js","path":"/_RootFolder/_ChildFolder/_GrandchildScript.js","parent":{"path":"/_RootFolder/_ChildFolder","id":"00000000-0000-0000-0000-000000000000"},"isFolder":false,"hasChildren":false,"id":"00000000-0000-0000-0000-000000000000"}]}",
      "type": "text",
    },
  ],
}
`;

exports[`script-tree get-script-tree-children should get children of root folder 1`] = `
{
  "content": [
    {
      "text": "{"total":2,"items":[{"name":"_ChildFolder","path":"/_RootFolder/_ChildFolder","parent":{"path":"/_RootFolder","id":"00000000-0000-0000-0000-000000000000"},"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_ChildScript.js","path":"/_RootFolder/_ChildScript.js","parent":{"path":"/_RootFolder","id":"00000000-0000-0000-0000-000000000000"},"isFolder":false,"hasChildren":false,"id":"00000000-0000-0000-0000-000000000000"}]}",
      "type": "text",
    },
  ],
}
`;

exports[`script-tree get-script-tree-children should get empty children for non-folder script 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`script-tree get-script-tree-root should get root level script items 1`] = `
{
  "content": [
    {
      "text": "{"total":3,"items":[{"name":"_RootFolder","path":"/_RootFolder","parent":null,"isFolder":true,"hasChildren":true,"id":"00000000-0000-0000-0000-000000000000"},{"name":"Parent Folder","path":"/Parent Folder","parent":null,"isFolder":true,"hasChildren":false,"id":"00000000-0000-0000-0000-000000000000"},{"name":"_RootScript.js","path":"/_RootScript.js","parent":null,"isFolder":false,"hasChildren":false,"id":"00000000-0000-0000-0000-000000000000"}]}",
      "type": "text",
    },
  ],
}
`;
